File: web/src/types/node.ts
import type { OperationProgress } from './operation'

// 命名约定：Node 等响应类型用 snake_case 匹配 Go JSON wire format；
// NodeFilterBase 等前端内部类型用 camelCase，由 api/node.ts toNodeParams 在 API 边界翻译。

export interface Transport {
  network?: string
  security?: string
  [key: string]: unknown
}

export interface ProtocolOption {
  value: string
  label: string
}

export interface Node {
  id: number
  subscription_id: number
  name: string
  protocol: string
  protocol_label: string
  address: string
  port: number
  raw_url: string
  transport: Transport
  latency: number
  created_at: string
  updated_at: string
}

export interface NodeFilterBase {
  protocol?: string
  subscriptionId?: number
  keyword?: string
  latencyStatuses?: NodeLatencyStatus[]
}

export interface NodeListQuery extends NodeFilterBase {
  cursor?: string
  limit?: number
}

export type NodeLatencyStatus = 'pending' | 'available' | 'timeout'

export interface NodeListResponse {
  items: Node[]
  next_cursor: string
}

export interface NodeSpeedTestStatus {
  running: boolean
  progress?: OperationProgress
  error?: string
}

File: server/internal/model/node.go
package model

import (
	"time"

	"v2ray-server/pkg/types"
)

type Node struct {
	ID             uint            `json:"id" gorm:"primaryKey"`
	SubscriptionID uint            `json:"subscription_id" gorm:"index"`
	Name           string          `json:"name" gorm:"not null;index"`
	Protocol       string          `json:"protocol" gorm:"not null;index"`
	Address        string          `json:"address" gorm:"not null"`
	Port           int             `json:"port" gorm:"not null"`
	RawURL         string          `json:"raw_url" gorm:"type:text"`
	RawConfig      types.Map       `json:"-" gorm:"type:text"`
	Transport      types.Transport `json:"transport" gorm:"type:text"`
	Latency        int64           `json:"latency" gorm:"index"`
	LatencyRank    int             `json:"-" gorm:"index"`
	CreatedAt      time.Time       `json:"created_at"`
	UpdatedAt      time.Time       `json:"updated_at"`
}

func (n *Node) IdentityKey() string {
	parsed := &types.ParsedNode{Address: n.Address, Port: n.Port, Protocol: n.Protocol, RawConfig: n.RawConfig}
	return parsed.IdentityKey()
}

type NodeFilter struct {
	SubscriptionID  uint
	Protocol        string
	Keyword         string
	LatencyStatuses []string
	Cursor          string
	Limit           int
	ExcludeID       uint
}
